glib/src/convert.rs

```diff
@@ -0,0 +1,264 @@
+/// string.
+#[derive(thiserror::Error, Debug)]
+pub enum CvtError {
+    #[error(transparent)]
+    Convert(#[from] Error),
+    #[error("{source} at offset {offset}")]
+    IllegalSequence {
+        #[source]
+        source: Error,
+        offset: usize,
+    },
+}
+
+impl CvtError {
+    #[inline]
+    fn new(err: Error, bytes_read: usize) -> Self {
+        if err.kind::<ConvertError>() == Some(ConvertError::IllegalSequence) {
+            Self::IllegalSequence {
+                source: err,
+                offset: bytes_read,
+            }
+        } else {
+            err.into()
+        }
+    }
+}
+
+#[doc(alias = "g_convert")]
+pub fn convert(
+    str_: &[u8],
+    to_codeset: &str,
+    from_codeset: &str,
+) -> Result<(Slice<u8>, usize), CvtError> {
+    assert!(str_.len() <= isize::MAX as usize);
+    let mut bytes_read = 0;
+    let mut bytes_written = 0;
+    let mut error = ptr::null_mut();
+    let result = unsafe {
+        ffi::g_convert(
+            str_.as_ptr(),
+            str_.len() as isize,
+            to_codeset.to_glib_none().0,
+            from_codeset.to_glib_none().0,
+            &mut bytes_read,
+            &mut bytes_written,
+            &mut error,
+        )
+    };
+    if result.is_null() {
+        Err(CvtError::new(unsafe { from_glib_full(error) }, bytes_read))
+    } else {
+        let slice = unsafe { Slice::from_glib_full_num(result, bytes_written as usize) };
+        Ok((slice, bytes_read))
+    }
+}
+
+#[doc(alias = "g_convert_with_fallback")]
+pub fn convert_with_fallback(
+    str_: &[u8],
+    to_codeset: &str,
+    from_codeset: &str,
+    fallback: Option<&str>,
+) -> Result<(Slice<u8>, usize), CvtError> {
+    assert!(str_.len() <= isize::MAX as usize);
+    let mut bytes_read = 0;
+    let mut bytes_written = 0;
+    let mut error = ptr::null_mut();
+    let result = unsafe {
+        ffi::g_convert_with_fallback(
+            str_.as_ptr(),
+            str_.len() as isize,
+            to_codeset.to_glib_none().0,
+            from_codeset.to_glib_none().0,
+            fallback.to_glib_none().0,
+            &mut bytes_read,
+            &mut bytes_written,
+            &mut error,
+        )
+    };
+    if result.is_null() {
+        Err(CvtError::new(unsafe { from_glib_full(error) }, bytes_read))
+    } else {
+        let slice = unsafe { Slice::from_glib_full_num(result, bytes_written as usize) };
+        Ok((slice, bytes_read))
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// A wrapper for [`std::io::Error`] that can hold an offset into an input string.
+#[derive(thiserror::Error, Debug)]
+pub enum IConvError {
+    #[error(transparent)]
+    Error(#[from] io::Error),
+    #[error("{source} at offset {offset}")]
+    WithOffset {
+        #[source]
+        source: io::Error,
+        offset: usize,
+    },
+}
+
+#[derive(Debug)]
+#[repr(transparent)]
+#[doc(alias = "GIConv")]
+pub struct IConv(ffi::GIConv);
+
+unsafe impl Send for IConv {}
```

Comment:
All functions are taking `&self`, so is this actually also `Sync`? Otherwise having the conversion functions take a `&mut self` would seem useful.

Reply:
Reference is the man page, suggests it is not `Sync`: https://www.man7.org/linux/man-pages/man3/iconv.3.html#ATTRIBUTES

Reply:
Yeah, changed it